sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -93,7 +93,7 @@ sealed abstract class UserDefinedFunction {
 private[spark] case class SparkUserDefinedFunction(
     f: AnyRef,
     dataType: DataType,
-    inputSchemas: Seq[Option[ScalaReflection.Schema]],
```

Comment:
quite a lot of outside libraries use this constructor. see for example:
https://github.com/salesforce/TransmogrifAI/blob/master/features/src/main/scala/com/salesforce/op/features/FeatureSparkTypes.scala#L269

can you suggest a workaround? i see no easy way to go from DataType to ExpressionEncoder...

Reply:
Hi @koertkuipers , the example seems use `UserDefinedFunction` instead of `SparkUserDefinedFunction`. And since Spark v3.0, `UserDefinedFunction` has been changed from case class to abstract class,  see [SPARK-26216](https://issues.apache.org/jira/browse/SPARK-26216).

cc @cloud-fan

Reply:
hi @Ngone51 yes changing it to use SparkUserDefinedFunction instead of UserDefinedFunction was easy. but the change from inputSchemas to inputEncoders poses a bit of a challenge.

Reply:
It's not go from DataType to ExpressionEncoder, which is impossible. It's from `T <: FeatureType : TypeTag` to `ExpressionEncoder[T]`. I think you need to adjust `FeatureSparkTypes` and support creating `ExpressionEncoder`. This is the cost of relying on internal APIs.

Reply:
while its nice to see case classes supported, i was surprised to find that another feature of Encoder is not supported: Option to indicate a udf argument could be null. for example one could define a udf for `(Option[String], Int) => String` to explicitly handle the case where the first input argument is null (or should i say its now None).

Reply:
in my testing i also found the api now to be somewhat inconsistent/confusing. basically sometimes `CatalystTypeConverters.createToScalaConverter` is used and sometimes `ExpressionEncoder.fromRow`, depending solely on if the the argument is a top level struct or not. but `CatalystTypeConverters.createToScalaConverter` and `ExpressionEncoder.fromRow` behave very differently, leading to inconsistent application.

for example this (contrived) usage works:
```
case class Person(name: String, age: Option[Int])
Seq((1, Person("john", Some(55))), (2, Person("mary", None))).toDF("id", "person").withColumn("age", udf{ p: Person1 => p.age }.apply(col("person")))
```
but this does not:
```
Seq((1, Seq(Person("john", Some(55)))), (2, Seq(Person("mary", None)))).toDF("id", "persons").withColumn("ages", udf{ s: Seq[Person1] => s.map(_.age) }.apply(col("persons")))
```
and while Option works nicely in Person case class (and also in tuples) Option does not work in a simple Seq:
```
Seq(Seq(Some(1), None)).toDF.withColumn("value", udf{ s: Seq[Option[Int]] => s.map(_.map(_ + 1)) }.apply(col("value")) )
```
this inconsistency will be hard to understand.

finally let me give some background why i am a little nervous about this change...
spark udfs have been somewhat limited for a long time. no support for case class, tuples, options. so many libraries have worked around that by defining their own udfs on top on SparkUserDefinedFunction. we do this inhouse too. it is easy to do this with type classes thanks to the composability of inputSchemas.
so now you replaced inputSchemas with inputEncoders. but ExpressionEncoder and TypeTags are not composable. i do not see a way for us to build on top of this for our own inhouse udfs. so then the alternative for us is to abandon our inhouse udfs and start using spark's udfs again, which now support case classes and tuples, which is nice! but the inconsistency of the api and lack of support for option makes that currently not viable to me. i realize this is a spark internal api and this is entirely my own problem. but i thought it was worth pointing out because i suspect i am not the only one that has done this. i think this is one of the more typical workarounds people have done using spark (and i am aware of multiple implementations of this workaround).

sorry for the long posts(s)

Reply:
`inputSchemas` is composable but it contains too little information, that's why the Spark UDF was so limited before.

Our goal is to make Spark UDF powerful enough so that people don't need to use internal APIs to build inhouse UDFs. But you are right that the support is not completed. @Ngone51 Can you take a closer look and see how to make it completed?

BTW, if you do need to keep your inhouse UDFs for a while, there is a way to create `ExpressionEncoder` from `Seq[DataType]`, by calling `RowEncoder.apply`. It only supports standard Spark external types, i.e. `Row`, not `case class`, which is the same as older versions of Spark.

Reply:
>  @Ngone51 Can you take a closer look and see how to make it completed?

Yea, I'm looking into it.

Reply:
@cloud-fan thanks i will take a look at RowEncoder